math/sqrt_double.cpp

```diff
@@ -9,41 +9,54 @@
  */
 #include <cassert>
 #include <iostream>
+ // Added this so we can use cout
+using std::cout;
 
-/** Bisection method implemented for the function \f$x^2-a=0\f$
- * whose roots are \f$\pm\sqrt{a}\f$ and only the positive root is returned.
- */
+ /** Bisection method implemented for the function \f$x^2-a=0\f$
+  * whose roots are \f$\pm\sqrt{a}\f$ and only the positive root is returned.
+  */
 double Sqrt(double a) {
-    if (a > 0 && a < 1) {
-        return 1 / Sqrt(1 / a);
-    }
-    double l = 0, r = a;
-    /* Epsilon is the precision.
-    A great precision is
-    between 1e-7 and 1e-12.
-    double epsilon = 1e-12;
-    */
-    double epsilon = 1e-12;
-    while (l <= r) {
-        double mid = (l + r) / 2;
-        if (mid * mid > a) {
-            r = mid;
-        } else {
-            if (a - mid * mid < epsilon) {
-                return mid;
-            }
-            l = mid;
-        }
-    }
-    return -1;
+
+	if (a > 0 && a < 1) {
+		return 1 / Sqrt(1 / a);
+	}
+	double l = 0, r = a;
+	/* Epsilon is the precision.
+	A great precision is
+	between 1e-7 and 1e-12.
+	double epsilon = 1e-12;
+	*/
+	double epsilon = 1e-12;
+	while (l <= r) {
```

Comment:
Now that I am looking at the code, the correct way seems to be the following:
```c
while ((r - l) > (2.f * epsilon)) { // till the interval is greater than twice the limit
    if (std::fabs(a - mid * mid) < epsilon) {
        return mid;
    }

    double mid = (l + r) / 2.f;  // interval midpoint
    if (mid * mid > a) {
        r = mid;
    }
    else {
        l = mid;
    }
}
```

Reply:
Also, in C++11, there is a better way of accessing machine epsilon: [reference](https://en.cppreference.com/w/cpp/types/numeric_limits/epsilon)